OpenRA.Mods.Common/Traits/Air/Aircraft.cs

```diff
@@ -531,22 +530,52 @@ void INotifyBecomingIdle.OnBecomingIdle(Actor self)
 				OnBecomingIdle(self);
 		}
 
+		public bool CanRearmAt(Actor host)
+		{
+			return rearmable != null && rearmable.Info.RearmActors.Contains(host.Info.Name) && rearmable.RearmableAmmoPools.Any(p => !p.FullAmmo());
+		}
+
+		public bool CanRepairAt(Actor host)
+		{
+			return repairable != null && repairable.Info.RepairActors.Contains(host.Info.Name) && self.GetDamageState() != DamageState.Undamaged;
+		}
+
 		protected virtual void OnBecomingIdle(Actor self)
 		{
-			if (Info.VTOL && Info.LandWhenIdle)
+			var atLandAltitude = self.World.Map.DistanceAboveTerrain(CenterPosition) == Info.LandAltitude;
+			if (atLandAltitude)
+			{
+				var host = GetActorBelow();
+				if (host != null && (CanRearmAt(host) || CanRepairAt(host)))
+				{
+					self.QueueActivity(new ResupplyAircraft(self));
+					return;
+				}
+			}
+
+			if (!atLandAltitude && Info.VTOL && Info.LandWhenIdle)
 			{
 				if (Info.TurnToLand)
 					self.QueueActivity(new Turn(self, Info.InitialFacing));
 
 				self.QueueActivity(new HeliLand(self, true));
 			}
-			else if (!Info.CanHover && (Info.TakeOffOnResupply || ReservedActor == null || self.World.Map.DistanceAboveTerrain(CenterPosition) != Info.LandAltitude))
+			else if (!Info.CanHover && (Info.TakeOffOnResupply || ReservedActor == null || !atLandAltitude))
+			{
 				self.QueueActivity(new FlyCircle(self, -1, Info.IdleTurnSpeed > -1 ? Info.IdleTurnSpeed : TurnSpeed));
```

Comment:
I may be missing something, but this would skip the TakeOff activity, wouldn't it?

Reply:
Well, aircraft still take off automatically when a Fly activity triggers, but I guess you're right, it would be better for consistency to queue a TakeOff instead.